src/typechecker/context-reduction.lisp

```diff
@@ -268,11 +268,28 @@ Returns (VALUES deferred-preds retained-preds defaultable-preds)"
         :do (setf subs (compose-substitution-lists subs (fundep-entail% env expr-preds pred known-tyvars)))
         :finally (return subs)))
 
+(defun expand-pred-into-superclasses (env pred)
+  "This function finds the class in ENV associated with PRED and
+recursively appends superclass predicates, with appropriate type
+substitutions."
+  (declare (type environment env)
+           (type ty-predicate pred)
+           (values ty-predicate-list &optional))
+  (let* ((class (lookup-class env (ty-predicate-class pred)))
+         (subs (mapcan #'match
+                       (ty-predicate-types (ty-class-predicate class))
+                       (ty-predicate-types pred))))
+    (cons pred
+          (apply #'append (mapcar (lambda (p) (expand-pred-into-superclasses env (apply-substitution subs p)))
```

Comment:
use `loop :append` not `apply`

Reply:
fixed in [467687c](https://github.com/coalton-lang/coalton/pull/1315/commits/467687c7ce047c2aabfbcda9f49caa9a10a5218d)